### PR TITLE
Laboratory work #1, Zvarichuk Nikolay - 23FPL1

### PR DESCRIPTION
**Pull request by Niko-Zvar**

### CODE REVIEW

File: lab_1_classify_by_unigrams/main.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 """
 Lab 1.
 
-Language detection
+Language detection (...)
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:4]: Revert this change

File: seminars/practice_1_run_debug.py [hunk 2]
```diff
@@ -28,7 +28,7 @@
 # third_num = first_num / second_num
```
Comment (artyomtugaryov) [inline seminars/practice_1_run_debug.py:28]: Revert this file

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Unfortunately, there is not implementation of the first lab. Hope you'll done on time

File: lab_1_classify_by_unigrams/main.py [hunk 3]
```diff
@@ -20,7 +19,16 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned
     """
-
+    if isinstance(text, str):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:22]: revert the condition and return `None` as soon as possible

File: lab_1_classify_by_unigrams/main.py [hunk 4]
```diff
@@ -34,7 +42,20 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
 
     In case of corrupt input arguments, None is returned
     """
-
+    if isinstance(tokens, list):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:45]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 5]
```diff
@@ -34,7 +42,20 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
+        freq = {}
+        for i in tokens:
+            if isinstance(i, str):
+                if i in freq:
+                    continue
+                average_freq = tokens.count(i) / len(tokens)
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:51]: you can move it inside the next if and set to `freq[i]` without creating `average_freq` variable

File: lab_1_classify_by_unigrams/main.py [hunk 6]
```diff
@@ -49,6 +70,13 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
 
     In case of corrupt input arguments, None is returned
     """
+    tokens = tokenize(text)
+    freq = calculate_frequencies(tokens)
+    if freq is not None and tokens is not None and isinstance(language, str):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:75]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 7]
```diff
@@ -64,7 +92,14 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
 
     In case of corrupt input arguments, None is returned
     """
-
+    squa_diff = 0
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:95]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -84,6 +119,29 @@ def compare_profiles(
     In case of corrupt input arguments or lack of keys 'name' and
     'freq' in arguments, None is returned
     """
+    if isinstance(unknown_profile, dict) and isinstance(profile_to_compare, dict):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:122]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 9]
```diff
@@ -105,6 +163,19 @@ def detect_language(
 
     In case of corrupt input arguments, None is returned
     """
+    detect_1 = compare_profiles(unknown_profile, profile_1)
+    detect_2 = compare_profiles(unknown_profile, profile_2)
+    if isinstance(detect_1, float) and isinstance(detect_2, float):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:168]: the same

File: lab_1_classify_by_unigrams/start.py [hunk 10]
```diff
@@ -1,22 +1,27 @@
...
     Launches an implementation
     """
-    with open("assets/texts/en.txt", "r", encoding="utf-8") as file_to_read_en:
-        en_text = file_to_read_en.read()
-    with open("assets/texts/de.txt", "r", encoding="utf-8") as file_to_read_de:
-        de_text = file_to_read_de.read()
-    with open("assets/texts/unknown.txt", "r", encoding="utf-8") as file_to_read_unk:
-        unknown_text = file_to_read_unk.read()
-    result = None
+    from lab_1_classify_by_unigrams.main import create_language_profile
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/start.py:11]: move the import to the top of the file

File: lab_1_classify_by_unigrams/start.py [hunk 11]
```diff
@@ -1,22 +1,27 @@
...
+    unk_profile = create_language_profile('unknown', unknown_text)
+    from lab_1_classify_by_unigrams.main import detect_language
+    result = detect_language(unk_profile, en_profile, de_profile)
+    print(result)
     assert result, "Detection result is None"
 
-
 if __name__ == "__main__":
-    main()
+    language()
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/start.py:27]: do not change the name function

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_1_classify_by_unigrams/main.py (hunk 3 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py:22]: Why?

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py:75]: I can't. I have the necessary condition.

File: lab_1_classify_by_unigrams/main.py (hunk 7 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py:95]: I can't. Otherwise, Python has nothing to turn to.

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py:122]: I cannot understand.

File: lab_1_classify_by_unigrams/main.py (hunk 9 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py:168]: This is for simplicity and convenience.

File: lab_1_classify_by_unigrams/main.py (hunk 1 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py]: done

File: seminars/practice_1_run_debug.py (hunk 2 above)
Comment (Niko-Zvar) [reply seminars/practice_1_run_debug.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 12]
```diff
@@ -3,9 +3,11 @@
 
 Language detection
 """
-# pylint:disable=too-many-locals, unused-argument, unused-variable
 
 
+
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:8]: revert

File: lab_1_classify_by_unigrams/main.py (hunk 3 above)
Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:25]: It is a good style - we tried it on seminars https://medium.com/swlh/return-early-pattern-3d18a41bba8

File: lab_1_classify_by_unigrams/main.py (hunk 4 above)
Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:47]: agree, Artyom

File: lab_1_classify_by_unigrams/main.py [hunk 13]
```diff
@@ -34,7 +44,18 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
     In case of corrupt input arguments, None is returned
     """
-
+    if isinstance(tokens, list):
+        freq = {}
+        for i in tokens:
+            if isinstance(i, str) and len(i) == 1:
+                if i in freq:
+                    continue
+                if isinstance(i, str):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:53]: How does it possible if `i` is not string in this line?

File: lab_1_classify_by_unigrams/main.py [hunk 14]
```diff
@@ -34,7 +44,18 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
+                    freq[i] = tokens.count(i) / len(tokens)
+            else:
+                return None
+        return freq
+    return None
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:58]: It is very difficult to understand what is going on. Please simplify

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:75]: inverse the condition and return `None` as soon as posible

File: lab_1_classify_by_unigrams/main.py (hunk 7 above)
Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:94]: What? I mean inverse the condition!

File: lab_1_classify_by_unigrams/main.py [hunk 15]
```diff
@@ -64,7 +91,14 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
     In case of corrupt input arguments, None is returned
     """
-
+    squa_diff = 0
+    if len(actual) == len(predicted):
+        for i, x in enumerate(actual):
+            squa_diff += (x - predicted[i]) ** 2
+    else:
+        return None
+    mse = round((squa_diff/len(actual)), 4)
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:100]: do not create the variable - return `as is`

File: lab_1_classify_by_unigrams/main.py (hunk 9 above)
Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:169]: see previous comments

File: lab_1_classify_by_unigrams/start.py [hunk 16]
```diff
@@ -1,22 +1,31 @@
...
+        unknown_text = file_unk.read()
+    unk_profile = create_language_profile('unknown', unknown_text)
+    if unk_profile and en_profile and de_profile is not None:
+        result = detect_language(unk_profile, en_profile, de_profile)
+        print(result)
+    else:
+        result = None
     assert result, "Detection result is None"
 
-
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/start.py:20]: revert the line

File: lab_1_classify_by_unigrams/start.py [hunk 17]
```diff
@@ -1,22 +1,31 @@
 """
 Language detection starter
 """
-# pylint:disable=too-many-locals, unused-argument, unused-variable
+
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/start.py:4]: remove these empy lines and add a new one after `pylint disable`

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_1_classify_by_unigrams/start.py (hunk 17 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/start.py]: done, it seems

File: lab_1_classify_by_unigrams/start.py (hunk 16 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/start.py:20]: done

File: lab_1_classify_by_unigrams/main.py (hunk 15 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py:97]: done, it seems

File: lab_1_classify_by_unigrams/main.py (hunk 14 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 13 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py]: done, it seems

File: lab_1_classify_by_unigrams/main.py (hunk 12 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 18]
```diff
@@ -34,6 +42,18 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(tokens, list):
+        return None
+    freq = {}
+    for i in tokens:
+        if not isinstance(i, str):
+            return None
+        if len(i) != 1:
+            return None
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:52]: why do you need this?

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:74]: agree with Artyom

File: lab_1_classify_by_unigrams/main.py (hunk 15 above)
Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:97]: no

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_1_classify_by_unigrams/main.py (hunk 18 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py:52]: because of mypy

Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py:52]: the same goes for the following comment

Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py:52]: and I don't understand what "no" means in the latter

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py:74]: done, it seems

File: lab_1_classify_by_unigrams/main.py (hunk 4 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 3 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 5 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 7 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 9 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/start.py (hunk 10 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/start.py]: done

File: lab_1_classify_by_unigrams/start.py (hunk 11 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/start.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 15 above)
Comment (Niko-Zvar) [reply lab_1_classify_by_unigrams/main.py:97]: done, it seems
